src/allmydata/web/directory.py

```diff
@@ -82,33 +81,53 @@ def make_handler_for(node, client, parentnode=None, name=None):
         return DirectoryNodeHandler(client, node, parentnode, name)
     return UnknownNodeHandler(client, node, parentnode, name)
 
-class DirectoryNodeHandler(RenderMixin, rend.Page, ReplaceMeMixin):
-    addSlash = True
+
+class DirectoryNodeHandler(ReplaceMeMixin, Resource, object):
 
     def __init__(self, client, node, parentnode=None, name=None):
-        rend.Page.__init__(self)
+        super(DirectoryNodeHandler, self).__init__()
         self.client = client
         assert node
         self.node = node
         self.parentnode = parentnode
         self.name = name
         self._operations = client.get_web_service().get_operations()
 
-    def childFactory(self, ctx, name):
-        name = name.decode("utf-8")
+    def getChild(self, name, req):
+        """
+        Dynamically create a child for the given request and name
+        """
+        # trying to replicate what I have observed as Nevow behavior
+        # for these nodes, which is that a URI like
+        # "/uri/URI%3ADIR2%3Aj...vq/" (that is, with a trailing slash
+        # or no further children) renders "this" page
+        name = name.decode('utf8')
         if not name:
-            raise EmptyPathnameComponentError()
+            # replicating Nevow behavior that complains about "empty
+            # path segments" .. but twisted.web sends in "name=None"
+            # for a URL like "/foo/bar/" as well as "/foo//bar"
+            # (i.e. a trailing slash means "name=None" as well)
+            if b'//' in req.path:
+                raise EmptyPathnameComponentError(
+                    u"The webapi does not allow empty pathname components, i.e. a double slash",
+                )
+            return self
         d = self.node.get(name)
-        d.addBoth(self.got_child, ctx, name)
-        # got_child returns a handler resource: FileNodeHandler or
-        # DirectoryNodeHandler
+        d.addBoth(self._got_child, req, name)
         return d
 
-    def got_child(self, node_or_failure, ctx, name):
-        req = IRequest(ctx)
-        method = req.method
-        nonterminal = len(req.postpath) > 1
-        t = get_arg(req, "t", "").strip()
+    def _got_child(self, node_or_failure, req, name):
+        """
+        Callback when self.node.get has returned
+        """
+        # we used to determine if a request was 'terminal' by
+        # examining req.postpath .. but that's a Nevow-ism, so we
+        # determine if this is a "terminal" request using only Twisted
+        # Web APIs, which means finding where "name" is in the path
+        u = URL.from_text(unicode(req.path)).to_iri()
+        assert name in u.path, "Can't find our child-name in the path"
+        nonterminal = (u.path[-1] != name)
+        t = get_arg(req, "t", "").strip()  # XXX looking like MultiFormatResource..
```

Comment:
As far as `req.postpath` -- I put the prior code back, because yes Twisted has a `.postpath` as well. The behavior does seem slightly different from Nevow's, though (which is maybe why I did the other thing?) .. and I'm not sure why the original check is "`len(req.postpath) > 1`" instead of just checking if there's anything at all in `req.postpath` :(

Reply:
I wonder what happens with a request path like `/uri/foo/bar/foo`.  In this case, traversal at the "foo" child of the "uri" resource looks like it is "terminal" but it's not, it just happens to have the same name as the terminal segment.

Maybe the Nevow version didn't have this shortcoming because it was based on `postpath` length instead.

Reply:
I agree that `.postpath` seems like a more-stable thing to check. I will adjust the code to use that again -- what I mean is that the behavior between "Nevow's .postpath" and "Twisted Web's .postpath" seems different (still trying to understand when/how).

The other part of the comment was just saying that i don't really understand why the (original) check is `> 1` instead of `>= 1`

Reply:
Okay I've got a slightly more sane-looking check now.